src/uploader.c

```diff
@@ -2451,36 +2475,50 @@ static void prepare_upload_state(uv_work_t *work)
 
     state->encrypted_file_name = encrypted_file_name;
 
-    // Calculate deterministic file id from encrypted file name
-    char *file_id = calloc(FILE_ID_SIZE + 1, sizeof(char));
-    if (!file_id) {
+    // Get random index used for encryption
+    uint8_t *index = calloc(SHA256_DIGEST_SIZE + 1, sizeof(uint8_t));
+    if (state->index) {
+        index = str2hex(strlen(state->index), (char *)state->index);
+    } else {
+        random_buffer(index, SHA256_DIGEST_SIZE);
+    }
+
+    char *index_as_str = hex2str(SHA256_DIGEST_SIZE, index);
+    if (!index_as_str) {
         state->error_status = STORJ_MEMORY_ERROR;
         return;
     }
 
-    calculate_file_id(state->bucket_id, state->encrypted_file_name, file_id);
-
-    file_id[FILE_ID_SIZE] = '\0';
-    state->file_id = file_id;
+    state->index = index_as_str;
 
-    // Caculate the file encryption key based on file id
-    char *file_key = calloc(DETERMINISTIC_KEY_SIZE + 1, sizeof(char));
-    if (!file_key) {
+    // Caculate the file encryption key based on the index
+    char *key_as_str = calloc(DETERMINISTIC_KEY_SIZE + 1, sizeof(char));
```

Comment:
==27331== 65 bytes in 1 blocks are definitely lost in loss record 38 of 57
==27331==    at 0x4C2FB55: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27331==    by 0x415709: prepare_upload_state (uploader.c:2495)
==27331==    by 0x54EF2DE: ??? (in /usr/lib/x86_64-linux-gnu/libuv.so.1.0.0)
==27331==    by 0x54FCCB6: ??? (in /usr/lib/x86_64-linux-gnu/libuv.so.1.0.0)
==27331==    by 0x5A1B6B9: start_thread (pthread_create.c:333)
==27331==    by 0x5D3782C: clone (clone.S:109)
==27331==